components/inventory.ts

```diff
@@ -76,15 +100,130 @@ export function createInventory(
         return inventoryUserCache.get(profileId)!
     }
 
+    // Get user data to check on location progression
+    const userData = getUserData(profileId, gameVersion)
+
     // add all unlockables to player's inventory
     const allunlockables = getVersionedConfig<Unlockable[]>(
         "allunlockables",
         gameVersion,
         true,
     ).filter((u) => u.Type !== "location") // locations not in inventory
 
+    let unlockables: Unlockable[] = allunlockables
+
+    if (getFlag("enableMasteryProgression")) {
+        const packagedUnlocks: Map<string, boolean> = new Map()
+
+        const challengesUnlockables =
+            controller.challengeService.getChallengesUnlockables()
+
+        /**
+         * Separates unlockable types and lookup for progression level
+         * on unlockables that are locked behind mastery progression level
+         */
+        const [unlockedItems, otherItems]: [Unlockable[], Unlockable[]] =
+            allunlockables
+                // Sorts packages and evergreen gear wrappers first, so related unlockables can be targeted after
+                .sort((_, b) =>
+                    b.Type === "package" ||
+                    (b.Type === "evergreenmastery" && b.Properties.Unlocks)
+                        ? 1
+                        : -1,
+                )
+                .reduce(
+                    (acc, unlockable) => {
```

Comment:
I feel we should move this to a local function for clarity, it's too nested now.